src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/X509Pal.cs

```diff
@@ -83,6 +72,42 @@ private partial class AppleX509Pal : ManagedX509ExtensionProcessor, IX509Pal
                 throw new NotSupportedException(SR.NotSupported_KeyAlgorithm);
             }
 
+            private static TAlgorithm DecodeECPublicKey<TAlgorithm>(ICertificatePal? certificatePal, Func<SafeSecKeyRefHandle, TAlgorithm> factory)
+                where TAlgorithm : AsymmetricAlgorithm
+            {
+                const int errSecInvalidKeyRef = -67712;
+                const int errSecUnsupportedKeySize = -67735;
+
+                if (certificatePal is AppleCertificatePal applePal)
+                {
+                    SafeSecKeyRefHandle key = Interop.AppleCrypto.X509GetPublicKey(applePal.CertificateHandle);
+
+                    // If X509GetPublicKey uses the new SecCertificateCopyKey API it can return an invalid
+                    // key reference for unsupported algorithms. This currently happens for the BrainpoolP160r1
+                    // algorithm in the test suite (as of macOS Mojave Developer Preview 4).
+                    if (key.IsInvalid)
+                    {
+                        throw Interop.AppleCrypto.CreateExceptionForOSStatus(errSecInvalidKeyRef);
+                    }
+                    // EccGetKeySizeInBits can fail for two reasons. First, the Apple implementation has changed
+                    // and we receive values from API that were not previously handled. In that case the
+                    // implementation will need to be adjusted to handle these values. Second, we deliberately
+                    // return 0 from the native code to prevent hitting buggy API implementations in Apple code
+                    // later.
+                    if (Interop.AppleCrypto.EccGetKeySizeInBits(key) == 0)
+                    {
+                        key.Dispose();
+                        throw Interop.AppleCrypto.CreateExceptionForOSStatus(errSecUnsupportedKeySize);
+                    }
+
+                    return factory(key);
```

Comment:
Since this is effectively a tail-call, would it make more sense to just have the key the return value, and avoid the callback and the generic method?